### PR TITLE
Adding support for MSAL telemetry cache data.

### PR DESCRIPTION
**Pull request by trwalke**
Fix for https://github.com/AzureAD/microsoft-identity-web/issues/1900

Here is the implementation of the telemetry cache details for `TokenCacheNotificationArgs`. 
It appears to me that the `MsalDistributedTokenCacheAdapter` seems to be the correct place to make these changes but I want to confirm this with you @jmprieur and @jennyf19.

- Are there any other locations where I need to implement this logic?
- Are there any expectations for the customer to implement their own versions of `MsalAbstractTokenCacheProvider` or `MsalDistributedTokenCacheAdapter`
- If the above is correct, do you know of anyway to gather this information from the customers implementation? I am trying to avoid adding to much additional api service.

### CODE REVIEW

**Comment by trwalke** [comment]
Cant build until after MSAL release.

File: src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs
Comment (trwalke) [reply src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs:231]: I was thinking of putting it there but it did not seem like the right place based on what I saw in the class. 
"Set of properties that the token cache serialization implementations might use to optimize the cache."

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (jennyf19) [inline src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:207]: >telemetryData [](http://example.com/codeflow?start=28&length=13)

 I would rather not take this now, as we still have not figured out our client side telemetry story and would prefer to not have API changes which we cannot remove later.

File: src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs
Comment (jennyf19) [reply src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs:231]: would rather not add to the public API.

Comment (bgavrilMS) [reply src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs:231]: It's a fair point, I'd add it to the hints. The hints was always designed to be a property bag to allow extensibility without us having to create overloads.

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (bgavrilMS) [reply src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:207]: Are you objecting to how the data is collected or to what data is collected? I agree with the previous comment on the "HOW".

But on the WHY, we have agreed with E&D team to deliver this, in a meeting facilitated by @TimHannMSFT . The spec is [here](https://microsoft-my.sharepoint-df.com/:w:/p/nebharg/EWrRAFHR6VdGvx5i_-uCZiUBROlxzCXZzLdRUJ8ePTKkgw?e=bHx3KO)

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:192]: nit: `telemetryData?.CacheLevel = ...`

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:15]: You need to also update `MsalMemoryTokenCacheProvider` - the outcome can be only L1.

File: tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test/L1L2CacheTests.cs:227]: These tests are ok, but higher level tests are always better because they are easier to read and cover the E2E. Have a look at `CacheExtensionsTests`. 

It should be pretty easy to hook up MSAL to a setup with L1+L2 cache and to a setup with only L1 cache. We should make sure we have a test for when L1 cache is disabled (this is one of the options).

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
See comments

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (trwalke) [reply src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:192]: using `telemetryData?` on the left side of the `=` does not work for me.

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (trwalke) [reply src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:194]: I put it here because it is in the null check for the result. I figured if you did not get a value from the L2 cache it doesn't really count as a cache hit. Or does it?

I guess I should move the logic out of the null check for the memory cache.

File: tests/Microsoft.Identity.Web.Test/CacheExtensionsTests.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test/CacheExtensionsTests.cs:66]: Consider adding variations for:

- `disableL1Cache` is `false. 
- MsalMemoryCacheProvider
- no cache (i.e. use MSAL's internal cache)

**Review by bgavrilMS — APPROVED** [review]
Approving code changes, but please consider the PR blocked until we figure out the path forward with telemetry.

**Review by jennyf19 — CHANGES_REQUESTED** [review]
marking as blocked until the telemetry questions are sorted out.

**Comment by bgavrilMS** [comment]
@jennyf19  @jennyf19 - looks like our partners want to use `ITelemetryClient`. Do we need more clarifications from them?

File: src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs
Comment (jennyf19) [inline src/Microsoft.Identity.Web.TokenCache/MsalAbstractTokenCacheProvider.cs:11]: remove unneeded using statements

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (jennyf19) [inline src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:13]: remove using statements

**Comment by jennyf19** [comment]
> @jennyf19 @jennyf19 - looks like our partners want to use `ITelemetryClient`. Do we need more clarifications from them?

We are seeing much progress with OTel, so, at this point, we would not use ITelemetryClient. Will know more in another week or two.

File: tests/Microsoft.Identity.Web.Test/TestTelemetryClient.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test/TestTelemetryClient.cs:13]: We want to deprecate `ITelemetryClient`. Can you please re-write the test to not use this?

File: src/Microsoft.Identity.Web.TokenCache/CacheSerializerHints.cs
Comment (jmprieur) [inline src/Microsoft.Identity.Web.TokenCache/CacheSerializerHints.cs:7]: Is it still needed?

**Comment by jennyf19** [comment]
@trwalke do we have an ETA on wrapping this up? We are doing a release early next week. thanks.

**Comment by trwalke** [comment]
> @trwalke do we have an ETA on wrapping this up? We are doing a release early next week. thanks.

Targeting end of week or a day before. I encountered an issue with MSAL tests as a result of some of MSAL's internal static caching which is causing tests to fail on this branch. Go ahead and release. @jennyf19

File: tests/Microsoft.Identity.Web.Test/TestTelemetryClient.cs
Comment (trwalke) [reply tests/Microsoft.Identity.Web.Test/TestTelemetryClient.cs:13]: I cant seems to find an easy way to get the cache telemetry data without some kind of telemetry client. Maybe it is better to update the tests when the new telemetry client is released. Atleast this is able to validate its functionality for now.

File: src/Microsoft.Identity.Web.TokenCache/TokenCacheExtensions.cs
Comment (trwalke) [inline src/Microsoft.Identity.Web.TokenCache/TokenCacheExtensions.cs:23]: The MSAL tests need a way to reset this cache because when they run in a batch, they have the potential to collide and tests fail. Enabling tests to reset this fixed this issue.

File: src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web.TokenCache/Distributed/MsalDistributedTokenCacheAdapter.cs:13]: This using should not be needed.

File: tests/Microsoft.Identity.Web.Test.Common/Mocks/MockHttpMessageHandler.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test.Common/Mocks/MockHttpMessageHandler.cs:59]: Use `Task.FromResult(responseMessage)`

File: tests/Microsoft.Identity.Web.Test.Common/Mocks/MockHttpMessageHandler.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test.Common/Mocks/MockHttpMessageHandler.cs:50]: See the comments from GitHubActions around string comparison.

File: tests/Microsoft.Identity.Web.Test/CacheExtensionsTests.cs
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Web.Test/CacheExtensionsTests.cs:256]: Please resolve the GitHub Actions warnings. Id.Web uses nullable. Have a look at https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving

**Review by bgavrilMS — APPROVED** [review]

File: tests/Microsoft.Identity.Web.Test/TestTelemetryClient.cs
Comment (trwalke) [reply tests/Microsoft.Identity.Web.Test/TestTelemetryClient.cs:13]: CC @jennyf19  @jmprieur

File: src/Microsoft.Identity.Web.TokenCache/TokenCacheExtensions.cs
Comment (trwalke) [reply src/Microsoft.Identity.Web.TokenCache/TokenCacheExtensions.cs:23]: CC @jennyf19 @jmprieur

**Comment by bgavrilMS** [comment]
All telemetry questions are sorted out. Cache Level is required data point for MSAL - irrespective of telemetry solution.

`ITelemetryClient` is not being deprecated for now, so it will appear there as well and it's acceptable way to test for this datapoint.

**Comment by jennyf19** [comment]
@bgavrilMS can you make sure someone from the Id Web side of things approves PRs in the future before merging them? Especially ones opened a few months ago. It's hard to keep track of what is ready to review and what is not.
Thanks

cc: @jmprieur @JoshLozensky @westin-m

**Comment by jennyf19** [comment]
> All telemetry questions are sorted out. Cache Level is required data point for MSAL - irrespective of telemetry solution.
> 
> `ITelemetryClient` is not being deprecated for now, so it will appear there as well and it's acceptable way to test for this datapoint.

also, we don't want `ITelemetryClient` that's the solution we are going with.

**Pull request merged**
